### PR TITLE
Report builder tech debt

### PR DESCRIPTION
**Pull request by NoahCarnahan**
This PR refactors the report builder javascript. No user facing changes. The main impetus was a desire to move the view models to a static JS file and to get some basic tests running. 

Best reviewed commit-by-commit
@esoergel

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 17 errors:

* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 166 - Missing semicolon. (semi)
* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 235 - Expected '===' and instead saw '=='. (eqeqeq)
* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 176 - Missing semicolon. (semi)
* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 229 - 'django' is not defined. (no-undef)
* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 223 - Expected '===' and instead saw '=='. (eqeqeq)
* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 68 - Missing trailing comma. (comma-dangle)
* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 40 - Redundant Boolean call. (no-extra-boolean-cast)
* corehq/apps/userreports/static/userreports/spec/PropertyListItem.spec.js, line 5 - 'it' is not defined. (no-undef)
* corehq/apps/userreports/static/userreports/spec/PropertyListItem.spec.js, line 15 - 'it' is not defined. (no-undef)
* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 27 - 'ReportBuilder' is not defined. (no-undef)
* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 240 - Expected '!==' and instead saw '!='. (eqeqeq)
* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 230 - 'django' is not defined. (no-undef)
* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 239 - 'django' is not defined. (no-undef)
* corehq/apps/userreports/static/userreports/spec/PropertyListItem.spec.js, line 28 - 'it' is not defined. (no-undef)
* corehq/apps/userreports/static/userreports/spec/PropertyListItem.spec.js, line 1 - 'describe' is not defined. (no-undef)
* corehq/apps/userreports/static/userreports/spec/PropertyListItem.spec.js, line 21 - 'it' is not defined. (no-undef)
* corehq/apps/userreports/static/userreports/js/builder_view_models.js, line 228 - 'django' is not defined. (no-undef)

File: corehq/apps/userreports/templates/userreports/partials/report_builder_configure_report.html [hunk 1]
```diff
@@ -42,6 +42,15 @@
...
         }
 
+
+        var getOrDefault = function(options, key, default_) {
+            if (options[key] === undefined) {
+                return default_
+            } else {
+                return options[key]
+            }
+        };
```
Comment (esoergel) [inline corehq/apps/userreports/templates/userreports/partials/report_builder_configure_report.html]: :+1:

File: corehq/apps/userreports/templates/userreports/partials/report_builder_configure_report.html [hunk 2]
```diff
@@ -199,8 +199,8 @@
             this.analyticsAction = getOrDefault(options, 'analyticsAction', null);
             this.analyticsLabel = getOrDefault(options, 'analyticsLabel', '{{ report_type }}');
 
-            this.hasFormatCol = ko.observable(getOrDefault(options, 'hasFormatCol', true));
-            this.hasCalculationCol = ko.observable(getOrDefault(options, 'hasCalculationCol', false));
+            this.hasFormatCol = getOrDefault(options, 'hasFormatCol', true);
+            this.hasCalculationCol = getOrDefault(options, 'hasCalculationCol', false);
```
Comment (esoergel) [inline corehq/apps/userreports/templates/userreports/partials/report_builder_configure_report.html]: :+1:

File: corehq/apps/userreports/templates/userreports/partials/report_builder_configure_report.html [hunk 3]
```diff
@@ -17,6 +17,10 @@
 {% block js-inline %}{{ block.super }}
     <script type="text/javascript">
 
+        var REPORT_TYPE = '{{ report_type }}';
+        var INITIAL_FILTERS = {{ initial_filters|JSON }};
+        var INITIAL_COLUMNS = {{ initial_columns|JSON }};
+        var SOURCE_TYPE = {{ form.source_type|JSON }};
```
Comment (esoergel) [inline corehq/apps/userreports/templates/userreports/partials/report_builder_configure_report.html]: :100:

**Comment by esoergel** [comment]
:athletic_shoe: :x:

**Comment by esoergel** [comment]
:heart: :heart: :heart: 
Lotta great stuff in this PR.  `hqDefine`, moving js out of templates, separation of concerns, general simplifications, `django.gettext`.  @dimagi/js-team would love it.

File: corehq/apps/userreports/static/userreports/js/builder_view_models.js [hunk 4]
```diff
@@ -0,0 +1,274 @@
+/* global django, ReportBuilder */
+hqDefine('userreports/js/builder_view_models.js', function () {
+
+    var getOrDefault = function(options, key, default_) {
```
Comment (orangejenny) [inline corehq/apps/userreports/static/userreports/js/builder_view_models.js]: fwiw, this is similar to http://underscorejs.org/#defaults

**Comment by NoahCarnahan** [comment]
Thanks! I used the js guide for some guidance.

File: corehq/apps/userreports/static/userreports/js/builder_view_models.js (hunk 4 above)
Comment (NoahCarnahan) [inline corehq/apps/userreports/static/userreports/js/builder_view_models.js]: Ah, thanks. Didn't know about that function.

**Comment by benrudolph** [comment]
💯

File: corehq/apps/userreports/templates/userreports/reportbuilder/configure_report.html [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        var initialColumns = {{ initial_columns|JSON }};
+        var sourceType = {{ form.source_type|JSON }};
+        var propertyOptions = {{ property_options|JSON }};
+        ReportBuilder.IS_NON_NUMERIC_MAP = propertyOptions.reduce(function(map, option) {
+            map[option.id] = option.is_non_numeric;
+            return map
+        }, {});
+
+
+        var ConfigForm = hqImport('userreports/js/builder_view_models.js').ConfigForm;
```
Comment (dannyroberts) [inline corehq/apps/userreports/templates/userreports/reportbuilder/configure_report.html]: Nice. Feels a lot cleaner!

**Comment by biyeun** [comment]
:heart:

**Pull request merged**